core/material.js

```diff
@@ -1,26 +1,45 @@
 import { create_program_object, create_shader_object, gl } from './context';
 import { Transform, Render } from '../components';
 
+import { vertex } from './shaders';
+import { fragment } from './shaders';
+
 export class Material {
-  constructor() {
+  constructor(options) {
     this.uniforms = {};
     this.attribs = {};
+    this.features = [];
+    this.features_from_components(Array.from(options.components) || []);
+  }
+
+  features_from_components(components) {
+    this.features = components
+      .reduce((memo, component) => {
+        if (component.features.length > 0) {
+          return memo.concat(component.features);
+        } else {
+          return memo;
+        }
+      }, [])
+      .filter((feature, i, features) => features.indexOf(feature) === i);
+  }
+
+  add_feature(feature) {
+    if (this.features.indexOf(feature) === -1) {
+      this.features.push(feature);
```

Comment:
Yeah, it looks like `this.features` should be a `Set`. Unfortunately Sets don't have the `map` method, but that can be easily fixed with `Array.from()` in `vertex` and `fragment`.